## Exercise 4.6 Combining and Exporting Data
### This script contains the following points
#### 1. Importing the order_products_combined dataframe from .pkl format
#### 2. Confirming the dimensions of the imported dataframe
#### 3. Determine a suitable way to combine the orders_products_combined dataframe with your products data set.
#### 4. Confirm the results of the merge using the merge flag
#### 5. Export the newly created dataframe as orders_products_merged in a suitable format

In [32]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [33]:
# Create a pathfolder shortcut
path = r'C:\Users\Victoria\Desktop\Data Analytics Career Foundry\4.0 Python Fundamentals for Data Analysts\Instacart Basket Analysis'

#### 1. Importing the order_products_combined dataframe from .pkl format

In [34]:
# Import 'orders_products_combined.pkl' as df_ords_prods_com
df_ords_prods_com = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

#### 2. Confirming the dimensions of the imported dataframe

In [35]:
# Confirm the dimensions of df_ords_prods_com
df_ords_prods_com.shape

(32434489, 9)

##### Dimensions of export from previos notebook were (32434489, 9). They match!

#### 3. Determine a suitable way to combine the orders_products_combined dataframe with your products data set.

In [36]:
# First, import 'products_checked.csv' as df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

In [37]:
# Check the outputs of df_ords_prods_com
df_ords_prods_com.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2539329        1             1                       2                  8   
2   2539329        1             1                       2                  8   
3   2539329        1             1                       2                  8   
4   2539329        1             1                       2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  
0                     NaN         196                  1          0  
1                     NaN       14084                  2          0  
2                     NaN       12427                  3          0  
3                     NaN       26088                  4          0  
4                     NaN       26405                  5          0

In [38]:
#Check the output of df_prods
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [39]:
# Drop 'Unnamed: 0', this column keeps importing on all my csv files even though I don't import the index column. (???)
df_prods = df_prods.drop(columns = ['Unnamed: 0'])

In [40]:
# Confirming column drop
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [41]:
#Confirm dimesnsions of df_prods
df_prods.shape

(49672, 5)

In [42]:
# Common identifier column of two dataframes is 'product_id', I will use this column to merge on.
# As stated in the exercise, For your Instacart project, you’ll only be working with data sets that have a full merge rate, so you won’t need to worry about this or apply any changes to the merge you just completed (using how = 'inner').
df_merged_larger = df_ords_prods_com.merge(df_prods, on = 'product_id', indicator = True)

In [43]:
# Check output of df_merged_larger
df_merged_larger.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2398795        1             2                       3                  7   
2    473747        1             3                       3                 12   
3   2254736        1             4                       4                  7   
4    431534        1             5                       4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

#### 4. Confirm the results of the merge using the merge flag

In [44]:
# Check the merge results using the merge flag
df_merged_larger['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [45]:
# Check the merge type without overwriting data
pd.merge(df_ords_prods_com, df_prods, on = 'product_id', how = 'outer', indicator = True)

order_id  user_id  order_number  orders_day_of_the_week  \
0         2539329.0      1.0           1.0                     2.0   
1         2398795.0      1.0           2.0                     3.0   
2          473747.0      1.0           3.0                     3.0   
3         2254736.0      1.0           4.0                     4.0   
4          431534.0      1.0           5.0                     4.0   
...             ...      ...           ...                     ...   
32435065        NaN      NaN           NaN                     NaN   
32435066        NaN      NaN           NaN                     NaN   
32435067        NaN      NaN           NaN                     NaN   
32435068        NaN      NaN           NaN                     NaN   
32435069        NaN      NaN           NaN                     NaN   

          order_hour_of_day  days_since_prior_order  product_id  \
0                       8.0                     NaN         196   
1                       7.0                    15.0         196   
2                      12.0                    21.0         196   
3                       7.0                    29.0         196   
4                      15.0                    28.0         196   
...                     ...                     ...         ...   
32435065                NaN                     NaN       37703   
32435066                NaN                     NaN       43725   
32435067                NaN                     NaN       45971   
32435068                NaN                     NaN       46625   
32435069                NaN                     NaN       49540   

          add_to_cart_order  reordered  \
0                       1.0        0.0   
1                       1.0        1.0   
2                       1.0        1.0   
3                       1.0        1.0   
4                       1.0        1.0   
...                     ...        ...   
32435065                NaN        NaN   
32435066                NaN        NaN   
32435067                NaN        NaN   
32435068                NaN        NaN   
32435069                NaN        NaN   

                                               product_name  aisle_id  \
0                                                      Soda      77.0   
1                                                      Soda      77.0   
2                                                      Soda      77.0   
3                                                      Soda      77.0   
4                                                      Soda      77.0   
...                                                     ...       ...   
32435065  Ultra Sun Blossom Liquid 90 loads Fabric Enhan...      75.0   
32435066                               Sweetart Jelly Beans     100.0   
32435067                         12 Inch Taper Candle White     101.0   
32435068    Single Barrel Kentucky Straight Bourbon Whiskey      31.0   
32435069                            Pure Squeezed  Lemonade      31.0   

          department_id  prices      _merge  
0                   7.0     9.0        both  
1                   7.0     9.0        both  
2                   7.0     9.0        both  
3                   7.0     9.0        both  
4                   7.0     9.0        both  
...                 ...     ...         ...  
32435065           17.0    14.3  right_only  
32435066           21.0     8.1  right_only  
32435067           17.0     9.8  right_only  
32435068            7.0     1.7  right_only  
32435069            7.0     9.1  right_only  

[32435070 rows x 14 columns]

##### Don't actually have a full match, but the data has a full merge rate...the exercise stated not to worry about this.

In [46]:
# Check dimensions of new, merged dataframe
df_merged_larger.shape

(32404859, 14)

#### 5. Export the newly created dataframe as orders_products_merged in a suitable format
##### I'm going to create a new merged data set without the merge flag because I had to remove it from my export in the other notebook in order to merge the data sets in this notebook.
##### I received an error that the indicator column could not be created because it already existed in the other notebook....
##### I'll export both just in case I need both, I'd rather do that now instead of having to possibly come back and do it later.

In [47]:
# Create merged dataframe with 'no indicator', df_merged_larger_ni
df_merged_larger_ni = df_prods.merge(df_ords_prods_com, on = 'product_id')

In [48]:
# Confirm the shape, should have one less column than df_merged_larger
df_merged_larger_ni.shape

(32404859, 13)

##### Confirmed, now I will export both in .pkl format because they are very large

In [49]:
# Export df_merged_larger as a .pkl file
df_merged_larger.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [50]:
# Export df_merged_larger_ni as a .pkl file
df_merged_larger_ni.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_ni.pkl'))